notebook2/notebook/notebook.py

```diff
@@ -188,6 +198,131 @@ def external_url_for(path):
+        return None
+
+    state = container_statuses[0].state
+    rn = None
+    wt = None
+    tm = None
+    if state.running:
+        rn = {"started_at": state.running.started_at}
+
+    if state.waiting:
+        wt = {"reason": state.waiting.reason}
+
+    if state.terminated:
+        tm = {"exit_code": state.terminated.exit_code, "finished_at": state.terminated.finished_at,
+              "started_at": state.terminated.started_at, "reason": state.terminated.reason}
+
+    if rn is None and wt is None and tm is None:
+        return None
+
+    return {"running": rn, "terminated": tm, "waiting": wt}
+
+
+def read_conditions(conds):
+    """
+        Return the most recent status=="True" V1PodCondition or None
+
+        Parameters
+        ----------
+        conds : list[V1PodCondition]
+            https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1PodCondition.md
+
+    """
+    if conds is None:
+        return None
+
+    maxDate = None
+    maxCond = None
+    for condition in conds:
+        if maxDate is None:
+            maxCond = condition
+            maxDate = condition.last_transition_time
+            continue
+
+        if condition.last_transition_time > maxDate and condition.status == "True":
+            maxCond = condition
+            maxDate = condition.last_transition_time
+
+    return {
+        "message": maxCond.message,
+        "reason": maxCond.reason,
+        "status": maxCond.status,
+        "type": maxCond.type}
+
+
+def marshall_notebooks(pods = [], svc_status = None):
+    notebooks = []
+
+    for pod in pods:
+        notebook = {
+            'name': pod.metadata.labels['name'],
+            'pod_name': pod.metadata.name,
+            'svc_name': pod.metadata.labels['svc_name'],
+            'pod_status': pod.status.phase,
+            'creation_date': pod.metadata.creation_timestamp.strftime('%D'),
+            'jupyter_token': pod.metadata.labels['jupyter_token'],
+            'container_status': read_containers_status(pod.status.container_statuses),
+            'condition': read_conditions(pod.status.conditions)
+        }
+
+        notebook['url'] = f"/instance/{notebook['svc_name']}/?token={notebook['jupyter_token']}"
+
+        if svc_status is not None:
+            notebook['svc_status'] = svc_status
+        else:
+            notebook['svc_status'] = read_svc_status(notebook['svc_name'])
+
+        notebooks.append(notebook)
+
+    return notebooks
+
+
+def get_live_user_notebooks(user_id = None):
+    if user_id is None:
+        return None
+
+    pods = k8s.list_namespaced_pod(
+        namespace='default',
+        label_selector=f"user_id={user_id}",
+        timeout_seconds=KUBERNETES_TIMEOUT_IN_SECONDS).items
+
+    if len(pods) == 0:
+        return None
+
+    # Kubernetes service will reflect the change immediately, pod will not
+    notebooks = list(filter(lambda n: n['svc_status'] != 'Deleted', marshall_notebooks(pods)))
```

Comment:
We can also check the pod for the presence of the deletionTimestamp metadata:
```bash
# k get pod dk-test-544786cc69-6rw6h -o yaml >y1 && \
  k delete pod dk-test-544786cc69-6rw6h && \
  k get pod dk-test-544786cc69-6rw6h -o yaml >y2
# diff y1 y2
7a8,9
>   deletionGracePeriodSeconds: 30
>   deletionTimestamp: 2019-03-04T15:47:31Z
21c23
<   resourceVersion: "32634684"
---
>   resourceVersion: "32634891"
# head y2
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/limit-ranger: 'LimitRanger plugin set: cpu request for container
      dk-test'
  creationTimestamp: 2019-03-04T15:46:10Z
  deletionGracePeriodSeconds: 30
  deletionTimestamp: 2019-03-04T15:47:31Z
```

Reply:
Useful, thanks!